gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java

```diff
@@ -0,0 +1,554 @@
+  private final Map<List<String>, Partition> sourcePartitions;
+  private final Map<List<String>, Partition> targetPartitions;
+
+  /**
+   * Defines what should be done for partitions that exist in the target but are not compatible with the source.
+   */
+  public enum ExistingEntityPolicy {
+    /** Deregister target partition, delete its files, and create a new partition with correct values. */
+    REPLACE,
+    /** Abort copying of conflict table. */
+    ABORT
+  }
+
+  /**
+   * A container for the differences between desired and existing files.
+   */
+  @Builder
+  private static class DiffPathSet {
+    /** Desired files that don't exist on target */
+    @Singular(value = "copyFile") Collection<FileStatus> filesToCopy;
+    /** Files in target that are not desired */
+    @Singular(value = "deleteFile") Collection<Path> pathsToDelete;
+  }
+
+  /**
+   * Represents a source {@link FileStatus} and a {@link Path} destination.
+   */
+  @Data
+  private static class SourceAndDestination {
+    private final FileStatus source;
+    private final Path destination;
+  }
+
+  HiveCopyEntityHelper(HiveDataset dataset, CopyConfiguration configuration, FileSystem targetFs) throws IOException {
+
+    this.dataset = dataset;
+    this.configuration = configuration;
+    this.targetFs = targetFs;
+
+    this.relocateDataFiles =
+        Boolean.valueOf(this.dataset.properties.getProperty(RELOCATE_DATA_FILES_KEY, DEFAULT_RELOCATE_DATA_FILES));
+    this.targetTableRoot = this.dataset.properties.containsKey(COPY_TARGET_TABLE_ROOT) ?
+        Optional.of(resolvePath(this.dataset.properties.getProperty(COPY_TARGET_TABLE_ROOT))) : Optional.<Path>absent();
+
+    this.hiveRegProps = new HiveRegProps(new State(this.dataset.properties));
+    this.targetURI = Optional.fromNullable(this.dataset.properties.getProperty(TARGET_METASTORE_URI_KEY));
+    this.targetClientPool = HiveMetastoreClientPool.get(this.dataset.properties, this.targetURI);
+    this.targetDatabase = Optional.fromNullable(this.dataset.properties.getProperty(TARGET_DATABASE_KEY)).
+        or(this.dataset.table.getDbName());
+    this.existingEntityPolicy = ExistingEntityPolicy.valueOf(
+        this.dataset.properties.getProperty(EXISTING_ENTITY_POLICY_KEY, DEFAULT_EXISTING_ENTITY_POLICY).toUpperCase());
+
+    try(AutoReturnableObject<IMetaStoreClient> sourceClient = this.dataset.clientPool.getClient();
+        AutoReturnableObject<IMetaStoreClient> targetClient = this.targetClientPool.getClient()) {
+
+      if (targetClient.get().tableExists(targetDatabase, this.dataset.table.getTableName())) {
+        this.existingTargetTable =
+            Optional.of(new Table(targetClient.get().getTable(targetDatabase, this.dataset.table.getTableName())));
+      } else {
+        this.existingTargetTable = Optional.absent();
+      }
+
+      if (this.existingTargetTable.isPresent()) {
+        checkTableCompatibility(this.dataset.table, this.existingTargetTable.get());
```

Comment:
Don't want to do that right now, it seems that replacing a table is a riskier operation than replacing partitions.